### PR TITLE
Proxy dex for xMEX

### PR DESCRIPTION
**Pull request by dorin-iancu**
Also upgraded all contracts to latest framework version and used common code from original proxy dex where possible.

### CODE REVIEW

File: dex/router/src/contract.rs [hunk 1]
```diff
@@ -83,45 +82,31 @@ pub trait Router:
...
         let caller = self.blockchain().get_caller();
-
         if caller != owner {
             require!(
                 self.pair_creation_enabled().get(),
                 "Pair creation is disabled"
             );
         }
 
-        require!(first_token_id != second_token_id, "Identical tokens");
```
Comment (psorinionut) [inline dex/router/src/contract.rs:96]: Why did you remove this check?

File: dex/router/src/contract.rs [hunk 2]
```diff
@@ -83,45 +82,31 @@ pub trait Router:
...
         let mut total_fee_percent_requested = DEFAULT_TOTAL_FEE_PERCENT;
         let mut special_fee_percent_requested = DEFAULT_SPECIAL_FEE_PERCENT;
 
         if caller == owner {
-            if let Some(fee_percents_multi_arg) = opt_fee_percents.into_option() {
-                let fee_percents_tuple = fee_percents_multi_arg.into_tuple();
-                total_fee_percent_requested = fee_percents_tuple.0;
-                special_fee_percent_requested = fee_percents_tuple.1;
-
-                require!(
```
Comment (psorinionut) [inline dex/router/src/contract.rs:117]: Why was this removed?

File: dex/router/src/contract.rs (hunk 1 above)
Comment (dorin-iancu) [reply dex/router/src/contract.rs:96]: The check is done inside the pair init function. It's duplicated code.

File: dex/router/src/contract.rs (hunk 2 above)
Comment (dorin-iancu) [reply dex/router/src/contract.rs:117]: Same reason as above.

File: locked-asset/proxy-dex-xmex/src/lib.rs [hunk 3]
```diff
@@ -37,17 +37,18 @@ pub trait ProxyDexImpl:
...
-        self.energy_factory_address()
-            .set_if_empty(&energy_factory_address);
+        self.old_locked_token_id().set(old_locked_token_id);
+        self.old_factory_address().set(old_factory_address);
+        self.energy_factory_address().set(energy_factory_address);
+        self.set_foundation_address(foundation_address);
+        self.set_lp_lock_epochs(lp_lock_epochs);
     }
 
     #[endpoint]
```
[the remark this answers is not in the document]
Comment (dorin-iancu) [reply locked-asset/proxy-dex-xmex/src/lib.rs]: People riot whenever I rename things, I'll leave it as it is.

File: locked-asset/proxy-dex-xmex/src/proxy_interactions/proxy_pair.rs [hunk 4]
```diff
@@ -45,41 +103,108 @@ pub trait ProxyPairModule:
...
+            other_payments: payments,
+            input_token_refs,
+            asset_token_id,
+            add_liq_result,
+        };
+
+        self.add_liq_common(add_liq_common_args)
+    }
+
+    fn add_liq_common<'a>(&self, args: AddLiqCommonArgs<Self::Api>) -> AddLiqResultType<Self::Api> {
```
[the remark this answers is not in the document]
Comment (dorin-iancu) [reply locked-asset/proxy-dex-xmex/src/proxy_interactions/proxy_pair.rs:178]: For the `AddLiqCommonArgs`. Compiler is smart enough to deduce it's for that argument, so I don't have to specify if there as well.

File: locked-asset/proxy-dex-xmex/src/proxy_interactions/proxy_pair.rs [hunk 5]
```diff
@@ -45,41 +103,108 @@ pub trait ProxyPairModule:
...
+            other_payments: payments,
+            input_token_refs,
+            asset_token_id,
+            add_liq_result,
+        };
+
+        self.add_liq_common(add_liq_common_args)
+    }
+
+    fn add_initial_liq_proxy(
```
[the remark this answers is not in the document]
Comment (dorin-iancu) [reply locked-asset/proxy-dex-xmex/src/proxy_interactions/proxy_pair.rs:132]: Meh, I'll leave it as it is for now, I don't quite get what you mean.

**Review by CostinCarabas — APPROVED** [review]

**Review by psorinionut — APPROVED** [review]
